packages/vats/src/vat-bank.js

```diff
@@ -0,0 +1,269 @@
+// @ts-check
+import { assert, details as X } from '@agoric/assert';
+import { amountMath, MathKind } from '@agoric/ertp';
+import { E } from '@agoric/eventual-send';
+import { Far } from '@agoric/marshal';
+import { makeNotifierKit, makeSubscriptionKit } from '@agoric/notifier';
+import { makeStore } from '@agoric/store';
+
+import { makeVirtualPurse } from './virtual-purse';
+
+import '@agoric/notifier/exported';
+
+/**
+ * @typedef {import('./virtual-purse').VirtualPurseController} VirtualPurseController
+ * @typedef {ReturnType<typeof makeVirtualPurse>} VirtualPurse
+ */
+
+/**
+ * @param {(obj: any) => Promise<any>} bankCall
+ * @param {string} denom
+ * @param {Brand} brand
+ * @param {string} address
+ * @param {Notifier<Amount>} balanceNotifier
+ * @returns {VirtualPurseController}
+ */
+const makePurseController = (
+  bankCall,
+  denom,
+  brand,
+  address,
+  balanceNotifier,
+) =>
+  harden({
+    async *getBalances(b) {
+      assert.equal(b, brand);
+      let updateRecord = await balanceNotifier.getUpdateSince();
+      while (updateRecord.updateCount) {
+        yield updateRecord.value;
+        // eslint-disable-next-line no-await-in-loop
+        updateRecord = await balanceNotifier.getUpdateSince(
+          updateRecord.updateCount,
+        );
+      }
+      return updateRecord.value;
+    },
+    async pushAmount(amt) {
+      const value = amountMath.getValue(amt, brand);
+      return bankCall({
+        type: 'VPURSE_GIVE',
+        recipient: address,
+        denom,
+        amount: `${value}`,
+      });
+    },
+    async pullAmount(amt) {
+      const value = amountMath.getValue(amt, brand);
+      return bankCall({
+        type: 'VPURSE_GRAB',
+        sender: address,
+        denom,
+        amount: `${value}`,
+      });
+    },
+  });
+
+/**
+ * @typedef {Object} AssetIssuerKit
+ * @property {Mint} [mint]
+ * @property {Issuer} issuer
+ * @property {Brand} brand
+ */
+
+/**
+ * @typedef {AssetIssuerKit & { denom: string }} AssetRecord
+ */
+
+/**
+ * @typedef {Object} Bank
+ * @property {() => Subscription<AssetDescriptor>}
+ * getAssetSubscription Returns assets as they are added to the bank
+ * @property {(brand: Brand) => VirtualPurse} getPurse Find any existing vpurse (keyed by address and brand) or create a
+ * new one.
+ */
+
+export function buildRootObject(_vatPowers) {
+  return Far('bankMaker', {
+    /**
+     * @param {import('./bridge').BridgeManager} bridgeManager
+     */
+    async makeBankManager(bridgeManager) {
+      /** @type {Store<Brand, AssetRecord>} */
+      const brandToAssetRecord = makeStore('brand');
```

Comment:
can this be Weak? It looks like we're never iterating

Reply:
It could be.  I will make it so.